applications/StructuralMechanicsApplication/custom_constraints/link_constraint.cpp

```diff
@@ -0,0 +1,335 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Máté Kelemen
+//
+
+// Project includes
+#include "custom_constraints/link_constraint.hpp" // LinkConstraint
+#include "includes/define.h"
+#include "includes/process_info.h"
+#include "includes/variables.h" // DISPLACEMENT_X, DISPLACEMENT_Y, DISPLACEMENT_Z
+#include "includes/serializer.h" // Serializer
+#include "includes/variables.h" // CONSTITUTIVE_MATRIX, INTERNAL_FORCES_VECTOR
+
+// STL includes
+#include <algorithm> // std::max_element, std::equal, std::transform
+
+
+namespace Kratos {
+
+
+struct LinkConstraint::Impl
+{
+    using ValueType = double;
+
+    static void ComputeRelationMatrix(LinkConstraint::MatrixType& rRelationMatrix,
+                                      LinkConstraint::MatrixType& rHessian,
+                                      LinkConstraint::VectorType& rConstraintGaps,
+                                      const std::array<ValueType,6>& rLastPositions,
+                                      const std::array<ValueType,6>& rLastDisplacements,
+                                      const unsigned Dimensions)
+    {
+        // Make sure we don't try to allocate an underflown unsigned integer.
+        KRATOS_ERROR_IF_NOT(Dimensions);
+
+        // Set output sizes.
+        rRelationMatrix.resize(1, 2 * Dimensions);
+        rHessian.resize(2 * Dimensions, 2 * Dimensions);
+        rConstraintGaps.resize(1);
+
+        // Initialize outputs.
+        rConstraintGaps[0] = 0.0;
+        std::fill(rRelationMatrix.data().begin(),
+                  rRelationMatrix.data().end(),
+                  static_cast<ValueType>(0));
+        std::fill(rHessian.data().begin(),
+                  rHessian.data().end(),
+                  static_cast<ValueType>(0));
+
+        // Compute constraint violation.
+        double current_norm = 0.0;
```

Comment:
This is not the usual norm, this is `current_norm_square`.

```suggestion
        double current_norm_square = 0.0;
```

Reply:
Well, you're thinking of the 2-norm but its square is also a norm.